File: clients/android/app/src/main/kotlin/com/lextures/android/core/network/ApiError.kt
package com.lextures.android.core.network

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive

sealed class ApiError : Exception() {
    data object InvalidResponse : ApiError() {
        private fun readResolve(): Any = InvalidResponse
        override val message: String = "Unexpected response from the server."
    }

    class HttpStatus(val code: Int, val apiMessage: String?) : ApiError() {
        override val message: String =
            if (!apiMessage.isNullOrBlank()) apiMessage else "Request failed (HTTP $code)."
    }

    class Transport(cause: Throwable) : ApiError() {
        override val message: String = cause.localizedMessage ?: cause.toString()
    }

    class Decoding(cause: Throwable) : ApiError() {
        override val message: String = cause.localizedMessage ?: cause.toString()
    }
}

/** Mirrors web `readApiErrorMessage` for common API error shapes. */
fun parseApiErrorMessage(body: String): String? {
    if (body.isBlank()) return null
    return runCatching {
        val json = Json.parseToJsonElement(body)
        if (json !is JsonObject) return@runCatching null
        listOf("message", "error", "detail").firstNotNullOfOrNull { key ->
            json[key]?.jsonPrimitive?.content?.takeIf { it.isNotBlank() }
        }
    }.getOrNull()
}

File: clients/ios/Lextures/Core/Networking/APIError.swift
import Foundation

enum APIError: LocalizedError {
    case invalidResponse
    case httpStatus(Int, message: String?)
    case decoding(Error)
    case transport(Error)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "Unexpected response from the server."
        case let .httpStatus(code, message):
            if let message, !message.isEmpty {
                return message
            }
            return "Request failed (HTTP \(code))."
        case let .decoding(error):
            return error.localizedDescription
        case let .transport(error):
            return error.localizedDescription
        }
    }
}

/// Mirrors web `readApiErrorMessage` for common API error shapes.
func parseAPIErrorMessage(from data: Data) -> String? {
    guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        return nil
    }
    if let error = json["error"] as? [String: Any],
       let message = error["message"] as? String,
       !message.isEmpty {
        return message
    }
    if let message = json["message"] as? String, !message.isEmpty {
        return message
    }
    if let error = json["error"] as? String, !error.isEmpty {
        return error
    }
    if let detail = json["detail"] as? String, !detail.isEmpty {
        return detail
    }
    return nil
}
